### PR TITLE
Implement URL Validation for Improved Data Access

### PR DESCRIPTION
**Pull request by manjirigunaji**
This pull request introduces a new module, validation_util.py, to handle URL validation within the National Water Model (NWM) utility scripts. The validation_util module includes a check_valid_urls function that validates URLs and filters out any invalid ones, reducing dependencies on tqdm. Additionally, URL validation functionality is to enhance the reliability of accessing data files in the project. 

#### Changes Made:
- Added a new module 'validation.py' containing functions for checking the validity of URLs.
- Integrated URL validation into the 'urlgen.py' module to ensure that only valid URLs are generated for data access.
- Implemented asynchronous URL validation using the 'gevent' library to improve performance and efficiency.

Tested various scenarios including valid URLs, invalid URLs, and edge cases to verify the robustness of the validation.

### CODE REVIEW

**Comment by jameshalgren** [comment]
Please post a few results from tests demonstrating the new function.

File: nwmurl/urlgennwm.py [hunk 1]
```diff
@@ -150,75 +189,15 @@ def select_lead_time(lead_time=None, default=None):
     9: "https://ciroh-nwm-zarr-copy.s3.amazonaws.com/national-water-model/",
 }
 
-retrospective_var_types = {
```
Comment (jameshalgren) [inline nwmurl/urlgennwm.py:153]: These deletions look accidental.

File: nwmurl/Readme_validation_testcase.txt [hunk 2]
```diff
@@ -0,0 +1,25 @@
+Here's an example of the test case output:
```
Comment (jameshalgren) [inline nwmurl/Readme_validation_testcase.txt:1]: Can you add the command line string to obtain this test?

**Review by jameshalgren — DISMISSED** [review]
Thank you for the PR! 
Please 
- [ ] correct the accidental deletions. 
- [ ] add the string to the test readme to obtain the test output indicated (also, please put this same information in the PR thread to facilitate review.)
- [ ] consider including a standard unit test for the validation script. 

Thanks! 

@karnesh Please watch for this updated PR.

**Review by karnesh — CHANGES_REQUESTED** [review]
Thank you for updating the PR! @manjirigunaji 

- Please clean up the code and remove same functions in urlgennwm.py and validation_util.py.
- Please post few results demonstrating the new function.
- It would be better to have a unit test instead of hard coded test - this can be either done now or in an another PR.

Thanks!

File: nwmurl/urlgennwm.py [hunk 3]
```diff
@@ -574,3 +618,91 @@ def generate_urls_operational(
...
+lead_time = [1,18]
+varinput = 1
+geoinput = 1
+runinput = 1
+urlbaseinput = 2
+meminput = 1
+generate_urls(start_date, end_date, fcst_cycle, lead_time, varinput, geoinput, runinput, urlbaseinput, meminput)
+# Example usage
+file_list = create_file_list(runinput, varinput, geoinput, meminput, start_date, end_date, fcst_cycle, urlbaseinput, lead_time)
+valid_files = check_valid_urls(file_list)
```
[the remark this answers is not in the document]
Comment (manjirigunaji) [reply nwmurl/urlgennwm.py:656]: Test
I have included a testing demo image that could be beneficial.

[the remark this answers is not in the document]
Comment (manjirigunaji) [reply nwmurl/urlgennwm.py:656]: The above is an  image  that is the filenamelist.txt.
image

**Review by karnesh — APPROVED** [review]
Thanks for updating the PR @manjirigunaji. Please change the hard coded test to a unit test in another PR. @sepehrkrz Please feel free to merge the PR.

**Review by karnesh — APPROVED** [review]
It looks good and this PR can be merged.

**Review by arpita0911patel — APPROVED** [review]

**Comment by jameshalgren** [comment]
This PR introduced some dependency collisions. Adding an issue for additional unit tests handling that -- @karnesh
